execution_engine/src/storage/trie_store/operations/mod.rs

```diff
@@ -394,6 +479,238 @@ where
+                                new_affix.extend(Vec::<u8>::from(extension_affix));
+                                let new_extension: Trie<K, V> = Trie::Extension {
+                                    affix: new_affix.into(),
+                                    pointer,
+                                };
+                                let trie_key = Blake2bHash::new(&new_extension.to_bytes()?);
+                                new_elements.push((trie_key, new_extension))
+                            }
+                        }
+                    }
+                }
+            }
+            // The parent is a pointer block, and we are propagating a node or extension upwards. It
+            // is impossible to propagate a leaf upwards.  Reseat the thing we are propagating into
+            // the parent.
+            (Some((trie_key, _)), Trie::Node { mut pointer_block }) => {
+                let trie_node: Trie<K, V> = {
+                    pointer_block[idx as usize] = Some(Pointer::NodePointer(*trie_key));
+                    Trie::Node { pointer_block }
+                };
+                let trie_key = Blake2bHash::new(&trie_node.to_bytes()?);
+                new_elements.push((trie_key, trie_node))
+            }
+            // The parent is an extension, and we are outputting an extension.  Prepend the
+            // parent affix to affix of the output extension, mutating the output in place.
+            // This is the only mutate-in-place.
+            (
+                Some((
+                    trie_key,
+                    Trie::Extension {
+                        affix: child_affix,
+                        pointer,
+                    },
+                )),
+                Trie::Extension { affix, .. },
+            ) => {
+                let mut new_affix: Vec<u8> = Vec::<u8>::from(affix);
+                new_affix.extend(Vec::<u8>::from(child_affix.to_owned()));
+                *child_affix = new_affix.into();
+                *trie_key = {
+                    let new_extension: Trie<K, V> = Trie::Extension {
+                        affix: child_affix.to_owned(),
+                        pointer: pointer.to_owned(),
+                    };
+                    Blake2bHash::new(&new_extension.to_bytes()?)
+                }
+            }
+            // The parent is an extension and the new element is a pointer block.
+            // The next element we add will be an extension to the pointer block we are going to
+            // add.
+            (Some((trie_key, Trie::Node { .. })), Trie::Extension { affix, mut pointer }) => {
+                pointer = Pointer::NodePointer(*trie_key);
+                let trie_extension = Trie::Extension { affix, pointer };
+                let trie_key = Blake2bHash::new(&trie_extension.to_bytes()?);
+                new_elements.push((trie_key, trie_extension))
+            }
+        }
+    }
+    let mut root_hash: Blake2bHash = root.to_owned();
```

Comment:
`new_elements` is never empty, is it? It would be clearer to not make this mutable then:
```suggestion
    let root_hash = new_elements.last().expect("new_elements can't be empty").0;
```

Reply:
Fixed https://github.com/CasperLabs/casper-node/pull/1155/commits/c2119df799123bac55303c1d81d5263fa26f9b9f